### PR TITLE
doc: provide context in glossary definitions

### PR DESCRIPTION
**Pull request by Qyriad**
Adds contextual information in the glossary for: derivation, store derivation, instantiate, realise, fixed-output derivation, Nix expression, and deriver.

# Motivation

This should hopefully make connecting the dots for some of these definitions a little easier, and make it clearer why these terms are relevant in the first place.

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by Lunaphied** [comment]
@zmitchell can you give this a look?

**Comment by zmitchell** [comment]
I'll give it a real review later tonight. I think a word could be tweaked here and there but otherwise it looks good to me.

File: doc/manual/src/glossary.md [hunk 1]
```diff
@@ -3,17 +3,24 @@
...
 
   A description of a build task. The result of a derivation is a
-  store object. Derivations are typically specified in Nix expressions
+  store object. Derivations created by Nix expressions are specified
   using the [`derivation` primitive](./language/derivations.md). These are
   translated into low-level *store derivations* (implicitly by
   `nix-env` and `nix-build`, or explicitly by `nix-instantiate`).
 
+  Package definitions, such as in [nixpkgs], are written using one or
+  more derivations.
```
Comment (zmitchell) [inline doc/manual/src/glossary.md:12]:
```suggestion
  more derivations, typically with the use of helper functions such as `mkDerivation`.
```

File: doc/manual/src/glossary.md [hunk 2]
```diff
@@ -3,17 +3,24 @@
...
   [derivation]: #gloss-derivation
+  [nixpkgs]: https://github.com/NixOS/nixpkgs
 
 - [store derivation]{#gloss-store-derivation}
 
   A [derivation] represented as a `.drv` file in the [store].
-  It has a [store path], like any [store object].
+  It has a [store path], like any [store object]. It is the
+  [instantiated][instantiate] form of a derivation. These are built
+  into *outputs* (implicitly by `nix-env` and `nix-build`, or explicitly
```
Comment (zmitchell) [inline doc/manual/src/glossary.md:22]:
```suggestion
  into *outputs* (implicitly by `nix-build`, or explicitly
```
We should refrain from mentioning `nix-env` as it's outdated and @fricklerhandwerk and the docs team have put in some good effort to remove it from official materials.

File: doc/manual/src/glossary.md [hunk 3]
```diff
@@ -25,7 +32,8 @@
 
   Translate a [derivation] into a [store derivation].
 
-  See [`nix-instantiate`](./command-ref/nix-instantiate.md).
+  See [`nix-instantiate`](./command-ref/nix-instantiate.md), which translates
```
Comment (zmitchell) [inline doc/manual/src/glossary.md:35]: I can't make a multi-line suggestion, but I would rephrase this as something like "...which produces store derivations from Nix expressions containing derivations". It's not crystal clear what `translate` means in this context.

File: doc/manual/src/glossary.md [hunk 4]
```diff
@@ -56,6 +64,11 @@
   A derivation which includes the
   [`outputHash`](./language/advanced-attributes.md#adv-attr-outputHash) attribute.
 
+  In package definitions such as in [nixpkgs], fixed-output derivations are
+  commonly created by by [fetchers].
```
Comment (zmitchell) [inline doc/manual/src/glossary.md:68]:
```suggestion
  commonly created by [fetchers].
```

File: doc/manual/src/glossary.md [hunk 5]
```diff
@@ -168,9 +181,10 @@
 
   A high-level description of software packages and compositions
   thereof. Deploying software using Nix entails writing Nix
-  expressions for your packages. Nix expressions are translated to
-  derivations that are stored in the Nix store. These derivations can
-  then be built.
+  expressions for your packages. Nix expressions specify [derviations],
+  which are [instantiated] and stored in the Nix store.
```
Comment (zmitchell) [inline doc/manual/src/glossary.md:185]:
```suggestion
  which are [instantiated] into Nix store paths.
```

Comment (zmitchell) [inline doc/manual/src/glossary.md:185]:
```suggestion
  which are [instantiated] as Nix store paths.
```

Comment (zmitchell) [reply doc/manual/src/glossary.md:185]: I can't decide which of these two I like better (or is more correct)

File: doc/manual/src/glossary.md [hunk 6]
```diff
@@ -168,9 +181,10 @@
...
+  These derivations can then be [realised][realise] (built), and produce
```
Comment (zmitchell) [inline doc/manual/src/glossary.md:186]: Multiline suggestion again: "... , to produce the derivation [outputs][output]".

**Review by zmitchell — CHANGES_REQUESTED** [review]
Ultimately @fricklerhandwerk has the actual say on this, but aside from some minor rephrasing in a couple of places it looks good to me.

File: doc/manual/src/glossary.md (hunk 5 above)
Comment (Qyriad) [reply doc/manual/src/glossary.md:185]: I don't know which is more correct, but the second is definitely less obvious. To the point that this along with your other suggestion below ("instantiated as Nix store paths") are connections that I myself hadn't fully made until now!

My only concern with this wording is that "Nix store path" is also its own jargon, and they're instantiated to a specific *kind* of Nix store path. Maybe "which are [instantiated] into the Nix store as store derivations"? Not sure if that's too much repetition of the same jargon.

Comment (Qyriad) [reply doc/manual/src/glossary.md:185]: Same as above

File: doc/manual/src/glossary.md (hunk 3 above)
Comment (zmitchell) [reply doc/manual/src/glossary.md:35]: That sounds fine to me

File: doc/manual/src/glossary.md [hunk 7]
```diff
@@ -168,9 +181,10 @@
 
   A high-level description of software packages and compositions
   thereof. Deploying software using Nix entails writing Nix
-  expressions for your packages. Nix expressions are translated to
-  derivations that are stored in the Nix store. These derivations can
-  then be built.
+  expressions for your packages. Nix expressions specify [derviations],
```
Comment (zmitchell) [inline doc/manual/src/glossary.md:184]:
```suggestion
  expressions for your packages. Nix expressions specify [derivations],
```

File: doc/manual/src/glossary.md (hunk 5 above)
Comment (zmitchell) [reply doc/manual/src/glossary.md:185]: That also works for me, less jargon is probably better

**Comment by Qyriad** [comment]
Applied all those suggestions (thank you!) and squashed. I also fixed the \[derivation] link for the store entry which I noticed was broken.

File: doc/manual/src/glossary.md [hunk 8]
```diff
@@ -3,17 +3,24 @@
...
   using the [`derivation` primitive](./language/derivations.md). These are
   translated into low-level *store derivations* (implicitly by
-  `nix-env` and `nix-build`, or explicitly by `nix-instantiate`).
+  `nix-build`, or explicitly by `nix-instantiate`).
+
+  Package definitions, such as in [nixpkgs], are written using one or
+  more derivations, typically with the use of helper functions such as `mkDerivation`.
 
   [derivation]: #gloss-derivation
+  [nixpkgs]: https://github.com/NixOS/nixpkgs
```
Comment (fricklerhandwerk) [inline doc/manual/src/glossary.md:15]:
```suggestion

  [derivation]: #gloss-derivation
```

Same argument as in the other comment: This is the Nix manual, the Nixpkgs manual should be capable of explaining the role of `mkDerivation` well enough. We now have nix.dev providing an entry point for beginners, clarifying that Nixpkgs is built on top of Nix, with the Nix language tutorial clearly highlighting that `derivation` is a low-level primitive that you'll never see or use in practice. I understand the argument that backlinks are valuable for discoverability and such, but not in this case where we're still fighting beginner confusion over separation of concerns.

File: doc/manual/src/glossary.md [hunk 9]
```diff
@@ -23,9 +30,10 @@
 
 - [instantiate]{#gloss-instantiate}, instantiation
 
-  Translate a [derivation] into a [store derivation].
+  Save an evaluated [derivation] as a low-level [store derivation] in the Nix store.
```
Comment (fricklerhandwerk) [inline doc/manual/src/glossary.md:33]: Hm... since I'm still quite unsure about what a derivation at the language level *is*, I have a bit of a hard time qualifying a derivation as evaluated or not. We can argue about the `derivation` primop with arguments being evaluated, but that sentence was deliberately bare-bones to avoid opening a can of worms or introducing irrelevant detail.

I also doubt that qualifying a store derivation as "low-level" helps. I know that the first entry of this glossary and the `nix-build` manual page call store derivations "low-level", but that's just because I haven't gotten to it yet. They're only (relatively) low-level from the perspective of anything outside of the store, and still they're exposed and dealt with explicitly in all kinds of interactions. And in terms of sentence structure, adding it as an adjective can be read as that there are at least two types of store derivations, regular ones and "low-level" ones.

At the very least, let's avoid the "low-level":

```suggestion
  Save an evaluated [derivation] as a [store derivation] in the Nix store.
```

But I don't think it says something wrong or misleading as it was.

**Review by fricklerhandwerk — CHANGES_REQUESTED** [review]
Thanks @Qyriad for the improvements and attention to detail, I really like the clarifications!

Some suggestions on questionable wording, and a hard change request on casually referencing Nixpkgs. Sorry for being wordy today, just trying to explain background and motivation in an attempt to avoid frustration with what may appear to be needlessly picky.

Please follow the [formatting guidelines](https://nixos.org/manual/nix/stable/contributing/documentation#markdown-conventions).

For the future, as I hope to see a lot more of contributions from you, consider doing multiple PRs for unrelated changes, because many of those I would have merged instantly. I've made good experience with that in avoiding getting blocked myself.

File: doc/manual/src/glossary.md [hunk 10]
```diff
@@ -3,17 +3,24 @@
...
+  [nixpkgs]: https://github.com/NixOS/nixpkgs
 
 - [store derivation]{#gloss-store-derivation}
 
   A [derivation] represented as a `.drv` file in the [store].
-  It has a [store path], like any [store object].
+  It has a [store path], like any [store object]. It is the
+  [instantiated][instantiate] form of a derivation. These are built
+  into *outputs* (implicitly by `nix-build`, or explicitly
+  by `nix-store --realise`).
```
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:23]: Maybe you mean that store derivations are the basis for realisation? Adding something like that could work out. But for the sake of merging quickly, this can be a follow-up PR.

File: doc/manual/src/glossary.md [hunk 11]
```diff
@@ -168,9 +181,10 @@
 
   A high-level description of software packages and compositions
   thereof. Deploying software using Nix entails writing Nix
-  expressions for your packages. Nix expressions are translated to
-  derivations that are stored in the Nix store. These derivations can
-  then be built.
+  expressions for your packages. Nix expressions specify [derivations][derivation],
+  which are [instantiated][instantiate] into the Nix store as [store derivations][store derivation].
+  These derivations can then be [realised][realise] (built) to produce
```
Comment (fricklerhandwerk) [inline doc/manual/src/glossary.md:186]:
```suggestion
  These derivations can then be [realised][realise] to produce
```
Maybe I'm getting paranoid, but I've spent so much effort to pry a proper definition for realisation out of the code and everyone else's minds, that I'm careful not to throw around words that aren't well defined. To "build" is a colloquialism for "running the `builder` executable of a derivation`, and that's just one possible way of realising it.

File: doc/manual/src/glossary.md (hunk 9 above)
Comment (Qyriad) [reply doc/manual/src/glossary.md:33]: It was a [suggested effort](https://github.com/NixOS/nix/pull/9378#discussion_r1398322002) to avoid the word "translate". I agree that the fuzziness of what exactly a derivation *is* makes this wording awkward. I don't mind reverting this part if you prefer.

File: doc/manual/src/glossary.md [hunk 12]
```diff
@@ -56,6 +64,11 @@
   A derivation which includes the
   [`outputHash`](./language/advanced-attributes.md#adv-attr-outputHash) attribute.
 
+  In package definitions such as in [nixpkgs], fixed-output derivations are
+  commonly created by [fetchers].
+
+  [fetchers]: https://nixos.org/manual/nixpkgs/stable/#chap-pkgs-fetchers
+
```
[the remark this answers is not in the document]
Comment (Qyriad) [reply doc/manual/src/glossary.md:71]: It doesn't specifically have to be a reference to Nixpkgs, however as you point out in that linked section the builtin fetchers don't create fixed-output derivations (something I wasn't 100% sure of until reading that section, thank you for that!), so they can't be used to provide context on why a fixed-output derivation matters or where it might come up.

Perhaps it could be mentioned that fixed-output derivations allow network access in the sandbox? (If that's even true; I assume that it is but didn't include it here since I couldn't find anything else documenting that)

File: doc/manual/src/glossary.md [hunk 13]
```diff
@@ -3,17 +3,24 @@
 - [derivation]{#gloss-derivation}
 
   A description of a build task. The result of a derivation is a
-  store object. Derivations are typically specified in Nix expressions
+  store object. Derivations created by Nix expressions are specified
```
[the remark this answers is not in the document]
Comment (Qyriad) [reply doc/manual/src/glossary.md:6]: Hm, maybe simply "Nix expressions \[can(?)] specify derivations with the `derivation` primitive"?

**Comment by Qyriad** [comment]
Thanks for the review!

> Please follow the [formatting guidelines](https://nixos.org/manual/nix/stable/contributing/documentation#markdown-conventions).

Ah, thanks, I wasn't sure if it was outdated or didn't apply to this page, since other parts of the page aren't yet following it.

> consider doing multiple PRs for unrelated changes, because many of those I would have merged instantly.

To be clear, you'd've preferred that each entry get its own PR?

**Comment by fricklerhandwerk** [comment]
> To be clear, you'd've preferred that each entry get its own PR?

Yes. Too often seemingly small changes are surprisingly far-reaching. But usually they're not and can be merged on the fly without being blocked by things that need a bit of thought or discussion.

File: doc/manual/src/glossary.md (hunk 9 above)
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:33]: "Translate" is indeed not great. Produce x from y?

File: doc/manual/src/glossary.md (hunk 8 above)
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:15]: > package definitions

Yes -- no. I don't have the issue by @roberth handy, but essentially: What's even a package? There's no such concept to be found here, to the best of my knowledge.

Also, more practically, Nix can be used to produce any file system data, whatever the problem domain this may encode, and I'm trying to capture that in the manual by being agnostic to applications.

File: doc/manual/src/glossary.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:6]: What's wrong with the passive voice here, using Nix expressions as a tool?

File: doc/manual/src/glossary.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:71]: FODs are only okay-documented, and I suggest not trying to solve that right here. We can add bits and pieces independently until someone runs into the urge of adding or rewriting a section.

File: doc/manual/src/glossary.md (hunk 8 above)
Comment (Qyriad) [reply doc/manual/src/glossary.md:15]: Well yes, but I'm not suggesting changing the definition of derivation in terms of packages, just to mention that derivations are very often used for making package definitions with Nix. And yes derivations are very agnostic (which is great!), but Nix *is* "the purely functional package manager", and plenty of other parts of the Nix manual reference the concept of "packages". I don't think it'd be overly presumptuous to say that "packages" are, or at least were, the motivation for derivations to exist, even if they can do more than that.

File: doc/manual/src/glossary.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (Qyriad) [reply doc/manual/src/glossary.md:6]: Nothing is wrong with the passive voice here; just trying to work around the fact that "derivation" is not a rigidly-defined concept.

I'm fine with your second suggestion.

File: doc/manual/src/glossary.md (hunk 8 above)
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:15]: It's true packages are mentioned here and there, but that's essentially artifacts that I remove when I work on a section containing them. I'm trying to get rid of terms that instill the wrong ideas or are effectively meaningless in this context, e.g. "installation". Eventually I will suggest rewording the front page of the manual.

Comment (Lunaphied) [reply doc/manual/src/glossary.md:15]: Yes. Nix is cool and could be used for lots of things. Is it? No. Not by anyone sane. It's fundamentally designed at it's core at the moment to be connected to package management and that needs to be acknowledged and explained in the documentation. If you want to describe that it's more versatile that's fine, but fucking document what it actually does

Comment (Lunaphied) [reply doc/manual/src/glossary.md:15]: Packages don't just "install software" they're simply a device that copies a set of files to a standardized location and registers them with a database for management — optionally transforming them, such as building software, patches, etc. — even typical package managers can install fonts, config files, other resources. A derivation *is* essentially a package, Nix's value is that it's more dynamic about how packages are ultimately built and installed. But don't pretend it's not making packages.

**Comment by fricklerhandwerk** [comment]
@Qyriad is it okay for you if I merge the already agreed-upon changes via a separate PR (preserving authorship information)? Then we can leave this one open as a small diff to discuss.

**Comment by Qyriad** [comment]
This is a reasonable request, but I'm afraid I've gotten too burned out to continue this PR. Feel free to merge the changes you link with our without attribution. Sorry about that.

**Comment by fricklerhandwerk** [comment]
Applied suggestions as discussed. Thanks a lot for putting in the work and sorry for the long wait!

**Comment by bzm3r** [comment]
**What is a derivation?**

Sorry, I fell into a rabbit-hole. Maybe some of this is useful. There are some questions it raises for me. There might be some consequences to this PR. 

From the [thesis](https://edolstra.github.io/pubs/phd-thesis.pdf) on `pdf_pg 35`, `numbered_pg 27`: 

> The result of the function in this case is a *derivation*. This is Nix-speak for a component build action, which derives the component from its inputs. We perform a derivation by calling `stdenv.mkDerivation`. `mkDerivation` is a function provided by `stdenv` that builds a component from a set of attributes.

Note that in terms of language, [a derivation is an action](https://en.wiktionary.org/wiki/derivation), and its result is a [derivate/derivative](https://en.wiktionary.org/wiki/derivate).

Here, the definition of derivation is immediately followed by a comment on `mkDerivation` (revealing a define-by-operation character of the thesis?). But that would be a distraction from the definition of derivation, which depends ultimately on the term *component*. 

On `pdf_pg 57`/`numbered_pg 50` (Section 3.1 What is a component?):

> ...we deploy software (although in Chapter 9 on service deployment, we will also see some things being deployed that cannot be called software proper), but what are the units of deployment? That is, what do we call the smallest meaningful artifacts that are subject to deployment? It is quite customary in some communities to call these *packages*, but this term has the problem that it implies that the artifact is in packaged form. For instance, a .RPM file is a package, but what do we call the corresponding installed object? In RPM terminology, this is also a package, and so the term is ambiguous.

> In the Component-Based Software Engineering (CBSE) community, the term component is used for units of deployment. Like some other terms in computer science...this term is overloaded almost to the point of uselessness...I shall use the term anyway, because I believe that the essential aspect that should be captured by a definition of components is deployment.

(Why does the choice of "component" for a concept follow from the requirement that that its essential aspect is that "deployment"? Would the term "deployable" not make more sense then? What does it mean "to deploy"? More on this later, when Eelco gives better motivation in the form of: component as a relative of the terms composable/composition.)

The thesis ultimately says:

> The important aspects of this [component] definition are the following:
> 
> • A component is a unit of independent deployment. Since it is a unit, it does not have to support partial deployment—the component must be deployed in its entirety. Since it is independently deployable, it should not cause interference with its target environment, such as other components.
> 
> • Thus it must have explicit context dependencies only, since otherwise it would place requirements (such as component dependencies) on its target environment that are not formally specified by its deployment metadata.
> 
> • It has contractually specified interfaces, which enable it to be subject to third-party composition, i.e., it is possible for other people’s components to use the component.
> 
> • It has no externally observable state. Different uses of the component over time should not interfere with each other. Note that this does not mean that the code of the component must have no side effects, i.e., should be written in a purely functional language. It is perfectly fine for an execution involving the component, e.g., the classes and objects loaded from a component, to have side effects. But the component should not modify its on-disk representation or other static resources in a way that affects future uses of the component. This is exactly the case with Nix components: they can be entirely impure dynamically, but their static representation—the files on disk—never changes after they have been built.

**Observation 1:** A derivation produces a *deployable* component. The term "deployable" itself is not defined in the thesis. Perhaps the best way to think of it might be (given the above definition + the meanings of the term deploy/deployment/deployable): "containing in itself all the data required for its concretization ([deployment](https://en.wiktionary.org/wiki/deployment)), hence: deployable".

**Observation 2:** A component is not a package. A component is an atomic (indivisible into further parts), deployable unit of a composable system. It has the specific character that it should be *automatically composable* in order to produce a larger system. Furthermore, it should be composable in a well-regulated manner.    

**Question 1:** would it be fair to say that a derivation deploys a component?

**Question 2:** given all of the above: why does there exist the term "deriver"? A *derivation* is *the* deriver of a derivative. Should the flag `--deriver` be renamed simply to `--derivation`?

**Question 3:** why do we call a derivation a "description of a build task"? It *is* the build task?

**Question 4:** why do we say "store derivation", rather than "stored derivation" (a derivation that is an object in the store).
___________________________________________

Side Hole:

In Chapter 9 of the thesis, examples are given of items that are not "software, proper":

> This chapter shows that Nix extends naturally into the domain of service deployment. A software service is a set of processes running on one or more machines that cooperate to provide a useful facility to an end-user or to another software system. An example might be a bug tracking service, implemented through a web server running certain web applications and a back-end database storing persistent data. A service is generally implemented through a set of software components, static data files, dynamic state (such as databases and log files), and configuration files that tie all these together.

**Future/Side Question 1:** how is this not software?

**Future/Side Question 2:** how are services typically handled by package managers in other Linux distributions? Are they "packages"?

**Future/Side Question 3:** is a service a component (per the definitions stated above?). It is clearly *deployable*. Is it also composable with other services? It is certainly not atomic as it is composed of smaller components. That's fine, since the definition of atomicity is choice-dependent. We get to choose our "zoom-level" when thinking of a system/we get to select the resolution of our model system by selecting what the minimum resolvable elements are ("atoms"). (For related mathematical context, a keyword: "pointless set theory/topology".)

**Review by fricklerhandwerk — APPROVED** [review]

**Comment by fricklerhandwerk** [comment]
Hey @bzm3r,
This is a bit too much to process in a PR comment thread, so I suggest you make direct and independent change proposals for the items you think should be fixed.

A few general thoughts on this from my side:
1. The derivation glossary entry is somewhat outdated and I deliberately haven't touched it with the goal to rework it once the actual documentation converges to something that can be concisely summarised. The definition of derivation I like most at the moment is the one here: https://nix.dev/manual/nix/2.19/language/ ("description of how to transform the contents of files to produce new files" – one could also say "description of pure computations on files", but that sounds needlessly academic)
2. I find that Eelco nailed it with "component" for the result of a derivation: it's super generic and therefore not wrong but also not helpful.
3. In light of that I'm not a fan of "build" or "action", because Nix really has nothing particular to do with the traditional notion of "building software", except for its original motivation. In a sense, everything is software (definitely everything Nix touches), and therefore all we do with it is computation. Therefore I try to avoid that terminology and stick to either the names of primitives defined by the existing interfaces, or use appropriately generic terms.
4. It's not wrong per se to use the term "package", but it needs an agreed-upon or authoritative definition. Right now we don't have one, therefore I frown upon its use. I'd be fine with defining "package" equivalent to the thesis's "component", roughly as as @Lunaphied suggested: an immutable file system object with explicit references to other file system objects. But it turns out that's exactly a store object! At the store and language level, Nix doesn't care about more semantics or the inner structure of these store objects, all that happens in Nixpkgs, and to some extent `nix-env` and `nix-shell` and friends, which I argue would benefit from being part of Nixpkgs.
4. Changing established names has social costs, and bears the potential to produce confusion.

File: doc/manual/src/glossary.md [hunk 14]
```diff
@@ -3,17 +3,18 @@
 - [derivation]{#gloss-derivation}
 
   A description of a build task. The result of a derivation is a
-  store object. Derivations are typically specified in Nix expressions
+  store object. Derivations declared in Nix expressions are specified
   using the [`derivation` primitive](./language/derivations.md). These are
```
Comment (roberth) [inline doc/manual/src/glossary.md:7]: Lies!
- The primitive is #9701

Please ignore.
